src/drivers/osd/osd.cpp

```diff
@@ -0,0 +1,636 @@
+
+	for (int i = 0; i < 5; i++) {
+		ret |= add_character_to_screen(buf[i], pos_x + i, pos_y);
+	}
+
+	ret |= add_character_to_screen(7, pos_x + 5, pos_y); // mAh symbol
+
+	return ret;
+}
+
+int
+OSD::add_altitude_symbol(uint8_t pos_x, uint8_t pos_y)
+{
+	return add_character_to_screen(154, pos_x, pos_y);
+}
+
+int
+OSD::add_altitude(uint8_t pos_x, uint8_t pos_y)
+{
+	char buf[5];
+	int ret = PX4_OK;
+
+	sprintf(buf, "%4.2f", (double)_local_position_z);
+
+	for (int i = 0; i < 5; i++) {
+		ret |= add_character_to_screen(buf[i], pos_x + i, pos_y);
+	}
+
+	ret |= add_character_to_screen('m', pos_x + 5, pos_y);
+
+	return ret;
+}
+
+int
+OSD::add_flighttime_symbol(uint8_t pos_x, uint8_t pos_y)
+{
+	return add_character_to_screen(112, pos_x, pos_y);
+}
+
+int
+OSD::add_flighttime(float flight_time, uint8_t pos_x, uint8_t pos_y)
+{
+	char buf[6];
+	int ret = PX4_OK;
+
+	sprintf(buf, "%5.1f", (double)flight_time);
+
+	for (int i = 0; i < 6; i++) {
+		ret |= add_character_to_screen(buf[i], pos_x + i, pos_y);
+	}
+
+	return ret;
+}
+
+int
+OSD::enable_screen()
+{
+	uint8_t data;
+	int ret = PX4_OK;
+
+	ret |= readRegister(0x00, &data, 1);
+	ret |= writeRegister(0x00, data | 0x48);
+
+	return ret;
+}
+
+int
+OSD::disable_screen()
+{
+	uint8_t data;
+	int ret = PX4_OK;
+
+	ret |= readRegister(0x00, &data, 1);
+	ret |= writeRegister(0x00, data & 0xF7);
+
+	return ret;
+}
+
+
+int
+OSD::update_topics()//TODO have an argument to choose what to update and return validity
+{
+	struct battery_status_s battery = {};
+	struct vehicle_local_position_s local_position = {};
+	struct vehicle_status_s vehicle_status = {};
+
+	bool updated = false;
+
+	/* update battery subscription */
+	orb_check(_battery_sub, &updated);
+
+	if (updated) {
+		orb_copy(ORB_ID(battery_status), _battery_sub, &battery);
+
+		if (battery.connected) {
+			_battery_voltage_filtered_v = battery.voltage_filtered_v;
+			_battery_discharge_mah = battery.discharged_mah;
+			_battery_valid = true;
+
+		} else {
+			_battery_valid = false;
+		}
+	}
+
+	/* update vehicle local position subscription */
+	orb_check(_local_position_sub, &updated);
+
+	if (updated) {
+		if (local_position.z_valid) {
+			_local_position_z = -local_position.z;
+			_local_position_valid = true;
+
+		} else {
+			_local_position_valid = false;
+		}
+	}
+
+	/* update vehicle status subscription */
+	orb_check(_vehicle_status_sub, &updated);
+
+	if (updated) {
+		if (vehicle_status.arming_state > 0 && _arming_state == 0) {
+			_arming_timestamp = hrt_absolute_time();
+		}
+
+		_arming_state = vehicle_status.arming_state;
+	}
+
+	return PX4_OK;
+}
+
+
+int
+OSD::update_screen()
+{
+	int ret = PX4_OK;
+
+	if (_battery_valid) {
+		ret |= add_battery_symbol(1, 1);
+		ret |= add_battery_info(2, 1);
+	}
+
+	if (_local_position_valid) {
+		ret |= add_altitude_symbol(1, 3);
+		ret |= add_altitude(2, 3);
+	}
+
+	if (_arming_state > 0) {
+		float flight_time_sec = static_cast<float>((hrt_absolute_time() - _arming_timestamp) / (1.0e6));
```

Comment:
Idea for later, this might be a useful thing to track more generally. Armed time, time in air, total distance flown, etc.